styx-scheduler-service/src/test/java/com/spotify/styx/SystemTest.java

```diff
@@ -114,6 +124,93 @@ public void removedEnabledWorkflowWontGetScheduled() throws Exception {
   }
 
   @Test
+  public void updatesNextNaturalTriggerWhenWFPartitioningChangesFromFinerToCoarser() throws Exception {
+    givenTheTimeIs("2016-03-14T15:30:00Z");
+    givenTheGlobalEnableFlagIs(true);
+    givenWorkflow(HOURLY_WORKFLOW);
+    givenWorkflowEnabledStateIs(HOURLY_WORKFLOW, true);
+
+    styxStarts();
+    timePasses(1, SECONDS);
+    awaitNumberOfDockerRuns(1);
+
+    WorkflowInstance workflowInstance = dockerRuns.get(0)._1;
+    RunSpec runSpec = dockerRuns.get(0)._2;
+    assertThat(workflowInstance.workflowId(), is(HOURLY_WORKFLOW.id()));
+    assertThat(runSpec, is(RunSpec.simple("busybox", "--hour", "2016-03-14T14")));
+
+    // this should store a new value for nextNaturalTrigger, 2016-03-15.
+    workflowChanges(DAILY_WORKFLOW);
+    timeJumps(1, DAYS);
+    timePasses(1, SECONDS);
+    awaitNumberOfDockerRuns(2);
+
+    workflowInstance = dockerRuns.get(1)._1;
+    runSpec = dockerRuns.get(1)._2;
+    assertThat(workflowInstance.workflowId(), is(DAILY_WORKFLOW.id()));
+    assertThat(runSpec, is(RunSpec.simple("busybox", "--hour", "2016-03-14")));
+  }
+
+  @Test
+  public void updatesNextNaturalTriggerWhenWFPartitioningChangesFromCoarserToFiner() throws Exception {
+    givenTheTimeIs("2016-03-14T15:30:00Z");
+    givenTheGlobalEnableFlagIs(true);
+    givenWorkflow(DAILY_WORKFLOW);
+    givenWorkflowEnabledStateIs(DAILY_WORKFLOW, true);
+
+    styxStarts();
+    timePasses(1, SECONDS);
+    awaitNumberOfDockerRuns(1);
+
+    WorkflowInstance workflowInstance = dockerRuns.get(0)._1;
+    RunSpec runSpec = dockerRuns.get(0)._2;
+    assertThat(workflowInstance.workflowId(), is(DAILY_WORKFLOW.id()));
+    assertThat(runSpec, is(RunSpec.simple("busybox", "--hour", "2016-03-13")));
+
+    // this should store a new value for nextNaturalTrigger, 2016-03-14T16.
+    workflowChanges(HOURLY_WORKFLOW);
+    timeJumps(1, HOURS);
+    timePasses(1, SECONDS);
+
+    awaitNumberOfDockerRuns(2);
+    workflowInstance = dockerRuns.get(1)._1;
+    runSpec = dockerRuns.get(1)._2;
+    assertThat(workflowInstance.workflowId(), is(DAILY_WORKFLOW.id()));
+    assertThat(runSpec, is(RunSpec.simple("busybox", "--hour", "2016-03-14T15")));
+  }
+
+  @Test
+  public void doesntUpdateNextNaturalTriggerWhenPartitioningDoesntChange() throws Exception {
+    givenTheTimeIs("2016-03-14T15:30:00Z");
+    givenTheGlobalEnableFlagIs(true);
+    givenWorkflow(DAILY_WORKFLOW);
+    givenWorkflowEnabledStateIs(DAILY_WORKFLOW, true);
+
+    styxStarts();
+    timePasses(1, SECONDS);
+    awaitNumberOfDockerRuns(1);
+
+    WorkflowInstance workflowInstance = dockerRuns.get(0)._1;
+    RunSpec runSpec = dockerRuns.get(0)._2;
+    assertThat(workflowInstance.workflowId(), is(DAILY_WORKFLOW.id()));
+    assertThat(runSpec, is(RunSpec.simple("busybox", "--hour", "2016-03-13")));
+
+    injectEvent(Event.started(workflowInstance));
+    injectEvent(Event.terminate(workflowInstance, 0));
+    Thread.sleep(1000);
```

Comment:
How about adding a way to await that a workflow instance completes? e.g. `awaitWorkflowInstanceCompletion(wfi)` that will wait until the state manager returns a null state'